File: Neetcode Questions/Linked List/lru-cache.cpp

// https://leetcode.com/problems/lru-cache/description/


class Node{
    public:
    int key, value;
    Node*next, *prev;
    Node(int k, int v){
        key = k;
        value = v;
        next = NULL;
        prev = NULL;
    }
};


class LRUCache {
public:
    int cap;
    Node*head = new Node(-1, -1);
    Node*tail = new Node(-1, -1);
    unordered_map<int, Node*> mp;

    LRUCache(int capacity) {
        cap = capacity;
        head->next = tail;
        tail->prev = head;
    }

    void deleteNode(Node*node){
        Node*delNext = node->next;
        Node*delPrev = node->prev;
        delPrev->next = delNext;
        delNext->prev = delPrev;
    }

    void addNode(Node*node){
        Node*temp = head->next;
        head->next = node;
        node->next = temp;
        temp->prev = node;
        node->prev = head;
    }
    
    int get(int key) {
        if(mp.find(key) != mp.end()){
            Node*temp = mp[key];
            int ans = temp->value;
            mp.erase(key);
            deleteNode(temp);
            addNode(temp);
            mp[key] = head->next;
            return ans;
        }
        return -1;
    }
    
    void put(int key, int value) {
        if(mp.find(key) != mp.end()){
            Node*temp = mp[key];
            mp.erase(key);
            deleteNode(temp);
        }

        if(mp.size() == cap){
            mp.erase(tail->prev->key);
            deleteNode(tail->prev);
        }

        addNode(new Node(key, value));
        mp[key] = head->next;   

    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: Neetcode-150-js/Linked List/lru-cache.js

// https://leetcode.com/problems/lru-cache/description/

class Node {
    constructor(key, value) {
        this.key = key;
        this.value = value;
        this.next = null;
        this.prev = null;
    }
}

class LRUCache {
    constructor(capacity) {
        this.size = capacity;
        this.head = new Node(-1, -1);
        this.tail = new Node(-1, -1);
        this.head.next = this.tail;
        this.tail.prev = this.head;
        this.map = new Map(); // Use JavaScript Map to store key-node pairs
    }

    deleteNode(node) {
        const { prev, next } = node;
        prev.next = next;
        next.prev = prev;
    }

    addNode(newNode) {
        const temp = this.head.next;
        this.head.next = newNode;
        newNode.next = temp;
        temp.prev = newNode;
        newNode.prev = this.head;
    }

    get(key) {
        if (this.map.has(key)) {
            const node = this.map.get(key);
            this.map.delete(key);
            this.deleteNode(node);
            this.addNode(node);
            this.map.set(key, this.head.next);
            return node.value;
        }
        return -1;
    }

    put(key, value) {
        if (this.map.has(key)) {
            const node = this.map.get(key);
            this.map.delete(key);
            this.deleteNode(node);
        }

        if (this.map.size === this.size) {
            const leastRecentNode = this.tail.prev;
            this.map.delete(leastRecentNode.key);
            this.deleteNode(leastRecentNode);
        }

        const newNode = new Node(key, value);
        this.addNode(newNode);
        this.map.set(key, this.head.next);
    }
}

/**
 * Example usage:
 * const obj = new LRUCache(capacity);
 * const param_1 = obj.get(key);
 * obj.put(key, value);
 */
